Implement the Locale claim for the OIDC connector

justin-slowik opened the pull request:
the .well-known/openid-configuration return of Dex was displaying that "locale" was a supported claim when in fact it was not.  My project requires locale, so I implemented the claim for the OIDC Connector and Storage methods.  

This is my first deep dive into Go so please let me know if i missed anything

srenatus commented on storage/sql/migrate.go:
```diff
@@ -98,6 +98,7 @@ var migrations = []migration{
 				claims_email text not null,
 				claims_email_verified boolean not null,
 				claims_groups bytea not null, -- JSON array of strings
+				claims_locale text not null,
```
:x: This won't work, I'm afraid. _Migrations_ are always run, even if the tables already exist because the service had been started before. So, to add a column, you'll need to `ALTER TABLE ... ADD COLUMN` in its own migration.

srenatus commented:
💯 Thank you for contributing! I've not reviewed this thoroughly yet, just pointed out one thing amiss wrt. migrations. I'll try to review this soon.

srenatus commented on examples/config-dev.yaml:
```diff
@@ -42,32 +42,65 @@ telemetry:
 #   level: "debug"
 #   format: "text" # can also be "json"
 
-# Default values shown below
```
Could this be left out, please? Improvements to the example config are of course welcome, but in a separate PR, please 😃

justin-slowik commented:
Hi @srenatus , I was just curious if you've had a chance to take another look at this PR since I addressed your comments.  Would love to get this in and then I could get working on zoneinfo next.  Thanks!

srenatus commented on storage/sql/crud.go:
```diff
@@ -152,14 +152,15 @@ func (c *conn) UpdateAuthRequest(id string, updater func(a storage.AuthRequest)
-				connector_id = $14, connector_data = $15,
-				expiry = $16
-			where id = $17;
+			    claims_locale = $14,
```
indentation looks a bit off here

justin-slowik replied:
weird, it looks normal on my IDE and gofmt ran clean.  Ill move it up a line and see if it aligns properly...

srenatus commented on storage/conformance/conformance.go:
```diff
@@ -98,6 +98,7 @@ func testAuthRequestCRUD(t *testing.T, s storage.Storage) {
 			Email:         "jane.doe@example.com",
 			EmailVerified: true,
 			Groups:        []string{"a", "b"},
+			Locale:        "en_US",
```
Can we have test cases where we there's _no_ locale information stored, too?

justin-slowik replied:
Removed locale in the secondary CRUD tests to verify that AuthRequest, AuthCode, and RefreshToken all function as intended.

justin-slowik commented on connector/oidc/oidc.go:
```diff
@@ -259,11 +259,14 @@ func (c *oidcConnector) HandleCallback(s connector.Scopes, r *http.Request) (ide
 		}
 	}
 
+	locale, _ := claims["locale"].(string)
```
I didn't write a unit test for it (yet), but in my ad-hoc testing it seemed to just ignore it.  ill add a test or 2

justin-slowik commented on storage/sql/migrate.go:
```diff
@@ -190,4 +190,16 @@ var migrations = []migration{
+	{
+		stmts: []string{`
+			alter table auth_request
+				add column claims_locale text not null default '';`,
```
actually just a habit from my current company's codebase.  I removed the not-null w/ default and seems to still behave fine, ill update accordingly

justin-slowik commented:
@srenatus in each of the conformance tests that utilize claims (testAuthCodeCRUD, testAuthRequestCRUD, and testRefreshTokenCRUD), I removed the locale claim from the secondary creates for each (a2 / refresh2 respectively) and verified that the tests still passed successfully. 

Long term it may be worth making a test / tests that exercises all claims one at a time to be more complete, but i think thats out of the scope of this PR.

justin-slowik commented:
@srenatus while waiting to hear back for any other review related changes, I created another branch to enable ZoneInfo.  I neglected to realize it would in turn update this PR.  So if thats cool w/ you I dont mind this being for both.  Otherwise I could break it off into a separate PR.  (my branch fo a branch was just easier to keep the SQL parameters more easily aligned).  Let me know what you prefer!

justin-slowik commented:
@srenatus Ignore my last comment/commit.  Accidentally pushed to the wrong branch.  Anything else that needs to be done here?

srenatus commented:
@justin-slowik I'd hope @JoelSpeed or @bonifaido would pick up reviewing and merging this. I have no bandwidth left at the moment, I'm afraid. (I could merge this, but I cannot find time to deal with any fallout, so I cannot actually merge this 😅)
